panic: sigs.k8s.io/cloud-provider-kind/pkg/controller.(*Controller).getRestConfig(

Copy from https://github.com/kubernetes/kubernetes/issues/139786#issuecomment-4743862652

```
I0616 10:52:34.020876    6132 app.go:90] FLAG: --enable-default-ingress="true"
I0616 10:52:34.020898    6132 app.go:90] FLAG: --enable-lb-port-mapping="false"
I0616 10:52:34.020900    6132 app.go:90] FLAG: --enable-log-dumping="true"
I0616 10:52:34.020901    6132 app.go:90] FLAG: --gateway-channel="standard"
I0616 10:52:34.020904    6132 app.go:90] FLAG: --help="false"
I0616 10:52:34.020905    6132 app.go:90] FLAG: --logs-dir="/logs/artifacts/loadbalancers"
I0616 10:52:34.020906    6132 app.go:90] FLAG: --verbosity="2"
I0616 10:52:34.021145    6132 app.go:161] **** Dumping load balancers logs to: /logs/artifacts/loadbalancers
E0616 10:59:04.924177    6132 controller.go:127] Failed to get kubeconfig for cluster kind: failed to get cluster internal kubeconfig: command "docker exec --privileged kind-control-plane cat /etc/kubernetes/admin.conf" failed with error: exit status 1
E0616 10:59:04.925809    6132 controller.go:147] "Failed to get internal kubeconfig" err="failed to get cluster internal kubeconfig: command \"docker exec --privileged kind-control-plane cat /etc/kubernetes/admin.conf\" failed with error: exit status 1" cluster="kind"
I0616 10:59:05.071295    6132 http.go:16] probe HTTP address https://127.0.0.1:37037
I0616 10:59:05.073986    6132 http.go:30] Failed to connect to HTTP address https://127.0.0.1:37037: Get "https://127.0.0.1:37037": EOF
I0616 10:59:05.074016    6132 controller.go:169] "Connected successfully" cluster="kind" host=""
I0616 10:59:05.081339    6132 app.go:113] Exiting: received signal
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x8 pc=0x1f269b2]

goroutine 1 [running]:
sigs.k8s.io/cloud-provider-kind/pkg/controller.(*Controller).getRestConfig(0x2bf7c3b13bf8, {0x2b194e0, 0x2bf7c3dac370}, {0x2bf7c39e8088, 0x4})
	/home/prow/go/src/sigs.k8s.io/cloud-provider-kind/pkg/controller/controller.go:176 +0x612
sigs.k8s.io/cloud-provider-kind/pkg/controller.(*Controller).Run(0x2bf7c3b75bf8, {0x2b194e0, 0x2bf7c3dac370})
	/home/prow/go/src/sigs.k8s.io/cloud-provider-kind/pkg/controller/controller.go:90 +0x347
sigs.k8s.io/cloud-provider-kind/cmd.runE(0x2bf7c3d90100?, {0x271db9e?, 0x4?, 0x271dba2?})
	/home/prow/go/src/sigs.k8s.io/cloud-provider-kind/cmd/app.go:217 +0xdbf
github.com/spf13/cobra.(*Command).execute(0x2bf7c3d1f808, {0x2bf7c3a260d0, 0x3, 0x3})
	/home/prow/go/pkg/mod/github.com/spf13/cobra@v1.10.2/command.go:1015 +0xb14
github.com/spf13/cobra.(*Command).ExecuteC(0x2bf7c3d1f808)
	/home/prow/go/pkg/mod/github.com/spf13/cobra@v1.10.2/command.go:1148 +0x465
github.com/spf13/cobra.(*Command).Execute(...)
	/home/prow/go/pkg/mod/github.com/spf13/cobra@v1.10.2/command.go:1071
sigs.k8s.io/cloud-provider-kind/cmd.Main()
	/home/prow/go/src/sigs.k8s.io/cloud-provider-kind/cmd/app.go:37 +0x18
main.main()
	/home/prow/go/src/sigs.k8s.io/cloud-provider-kind/main.go:8 +0xf
```

we need to fix this panic, I think the problem is we receive a signal to exit and the context cancellation does not handle it correctly

> I0616 10:59:05.074016    6132 controller.go:169] "Connected successfully" cluster="kind" host=""
I0616 10:59:05.081339    6132 app.go:113] Exiting: received signal
panic: runtime error: invalid memory address or nil pointer dereference

--- a/pkg/controller/controller.go
+++ b/pkg/controller/controller.go
@@ -171,23 +171,28 @@ func (c *Controller) getRestConfig(ctx context.Context, cluster string) (*rest.C
 		}
 	}
 
-	var config *rest.Config
-	switch host {
-	case internalConfig.Host:
-		config = internalConfig
+	return selectRestConfig(internalConfig, externalConfig, host)
+}
+
+// selectRestConfig returns the rest.Config whose Host matches the host that was
+// successfully probed. internalConfig or externalConfig may be nil when the
+// corresponding kubeconfig could not be obtained, so their Host is only
+// compared when they are set to avoid a nil pointer dereference.
+func selectRestConfig(internalConfig, externalConfig *rest.Config, host string) (*rest.Config, error) {
+	switch {
+	case internalConfig != nil && host == internalConfig.Host:
 		// the first cluster will give us the type of connectivity between
 		// cloud-provider-kind and the clusters and load balancer containers.
 		// In Linux or containerized cloud-provider-kind this will be direct.
 		once.Do(func() {
 			cpkconfig.DefaultConfig.ControlPlaneConnectivity = cpkconfig.Direct
 		})
-	case externalConfig.Host:
-		config = externalConfig
+		return internalConfig, nil
+	case externalConfig != nil && host == externalConfig.Host:
+		return externalConfig, nil
 	default:
 		return nil, fmt.Errorf("restConfig for host %s not avaliable", host)
 	}
-
-	return config, nil
 }
 
 // TODO: implement leader election to not have problems with multiple providers

--- a/pkg/controller/controller_test.go
+++ b/pkg/controller/controller_test.go
@@ -10,27 +10,74 @@ func Test_selectRestConfig(t *testing.T) {
 	internal := &rest.Config{Host: "https://internal:6443"}
 	external := &rest.Config{Host: "https://external:6443"}
 
-	// internal kubeconfig unavailable, only the external endpoint answered the
-	// probe: must not panic and must return the external config.
-	got, err := selectRestConfig(nil, external, external.Host)
-	if err != nil {
-		t.Fatalf("selectRestConfig() error = %v", err)
-	}
-	if got != external {
-		t.Errorf("selectRestConfig() = %v, want external config", got)
+	tests := []struct {
+		name           string
+		internalConfig *rest.Config
+		externalConfig *rest.Config
+		host           string
+		want           *rest.Config
+		wantErr        bool
+	}{
+		{
+			name:           "internal kubeconfig unavailable, external endpoint answered",
+			internalConfig: nil,
+			externalConfig: external,
+			host:           external.Host,
+			want:           external,
+		},
+		{
+			name:           "external kubeconfig unavailable, internal endpoint answered",
+			internalConfig: internal,
+			externalConfig: nil,
+			host:           internal.Host,
+			want:           internal,
+		},
+		{
+			name:           "both available, host matches internal",
+			internalConfig: internal,
+			externalConfig: external,
+			host:           internal.Host,
+			want:           internal,
+		},
+		{
+			name:           "both available, host matches external",
+			internalConfig: internal,
+			externalConfig: external,
+			host:           external.Host,
+			want:           external,
+		},
+		{
+			name:           "host matches neither available config",
+			internalConfig: internal,
+			externalConfig: external,
+			host:           "https://other:6443",
+			wantErr:        true,
+		},
+		{
+			name:           "only external available but host does not match",
+			internalConfig: nil,
+			externalConfig: external,
+			host:           internal.Host,
+			wantErr:        true,
+		},
+		{
+			name:           "both configs unavailable",
+			internalConfig: nil,
+			externalConfig: nil,
+			host:           internal.Host,
+			wantErr:        true,
+		},
 	}
 
-	// external kubeconfig unavailable, only the internal endpoint answered.
-	got, err = selectRestConfig(internal, nil, internal.Host)
-	if err != nil {
-		t.Fatalf("selectRestConfig() error = %v", err)
-	}
-	if got != internal {
-		t.Errorf("selectRestConfig() = %v, want internal config", got)
-	}
-
-	// host does not match any available config.
-	if _, err := selectRestConfig(nil, external, ""); err == nil {
-		t.Errorf("selectRestConfig() expected error for unmatched host, got nil")
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			got, err := selectRestConfig(tt.internalConfig, tt.externalConfig, tt.host)
+			if (err != nil) != tt.wantErr {
+				t.Fatalf("selectRestConfig() error = %v, wantErr %v", err, tt.wantErr)
+			}
+			if got != tt.want {
+				t.Errorf("selectRestConfig() = %v, want %v", got, tt.want)
+			}
+		})
 	}
 }

--- a/pkg/controller/controller_test.go
+++ b/pkg/controller/controller_test.go
@@ -1,0 +1,36 @@
+package controller
+
+import (
+	"testing"
+
+	"k8s.io/client-go/rest"
+)
+
+func Test_selectRestConfig(t *testing.T) {
+	internal := &rest.Config{Host: "https://internal:6443"}
+	external := &rest.Config{Host: "https://external:6443"}
+
+	// internal kubeconfig unavailable, only the external endpoint answered the
+	// probe: must not panic and must return the external config.
+	got, err := selectRestConfig(nil, external, external.Host)
+	if err != nil {
+		t.Fatalf("selectRestConfig() error = %v", err)
+	}
+	if got != external {
+		t.Errorf("selectRestConfig() = %v, want external config", got)
+	}
+
+	// external kubeconfig unavailable, only the internal endpoint answered.
+	got, err = selectRestConfig(internal, nil, internal.Host)
+	if err != nil {
+		t.Fatalf("selectRestConfig() error = %v", err)
+	}
+	if got != internal {
+		t.Errorf("selectRestConfig() = %v, want internal config", got)
+	}
+
+	// host does not match any available config.
+	if _, err := selectRestConfig(nil, external, ""); err == nil {
+		t.Errorf("selectRestConfig() expected error for unmatched host, got nil")
+	}
+}

